library/src/scala/deriving/Mirror.scala

```diff
@@ -49,4 +49,15 @@ object Mirror {
   type Of[T] = Mirror { type MirroredType = T; type MirroredMonoType = T ; type MirroredElemTypes <: Tuple }
   type ProductOf[T] = Mirror.Product { type MirroredType = T; type MirroredMonoType = T ; type MirroredElemTypes <: Tuple }
   type SumOf[T] = Mirror.Sum { type MirroredType = T; type MirroredMonoType = T; type MirroredElemTypes <: Tuple }
+
+  extension [T](p: ProductOf[T])
+    /** Create a new instance of type `T` with elements taken from product `a`. */
+    @annotation.experimental
+    def fromProductTyped[A <: scala.Product](a: A)(using m: ProductOf[A], ev: p.MirroredElemTypes =:= m.MirroredElemTypes): T =
+      p.fromProduct(a)
+
+    /** Create a new instance of type `T` with elements taken from tuple `t`. */
+    @annotation.experimental
+    def fromTuple(t: p.MirroredElemTypes): T =
```

Comment:
If `MirroredElemTypes` would be defined in `Mirror.Product` we could have defined this method directly in `Mirror.Product`.

@odersky was there a reason to put it where it is now? 
It seems that we could move it without breaking binary compatibility. Could we move it?

Reply:
Plausibly because you could define elem types in terms of your own HList instead of the standard Tuple, e.g. shapeless might want that?  I have no strong feeling about that (and I don't know how active shapeless 3 is).

Reply:
@milessabin how is this used in shapeless 3?

Reply:
From what I can see shapeless uses Tuples.

Reply:
Actually here is an example that shows why it is not defined in `Mirror`: https://github.com/lampepfl/dotty/blob/master/tests/pos/mirror-implicit-scope.scala#L12

Reply:
The current version LGTM